Add support for multiway union operation

iljakuklic opened the pull request:
Thought I could contribute some self-contained code that has arisen while working on another project.

It is support for the `multiway_union` set operation. The main difference between this and the existing union operation is that `multiway_union` takes a collection of iterators to be unionized (?) as opposed to just a pair of iterators.

iljakuklic committed:
Add support for multiway union operation

iljakuklic commented on src/sorted_iterator.rs:
```diff
@@ -54,6 +55,93 @@ impl<K: Ord, I: Iterator<Item = K>, J: Iterator<Item = K>> Iterator for Union<I,
+        self.bh.pop().map(|Peeked { h: cur, t: tail }| {
+            // Collect elements equivalent to the current element.
+            // This is done as a two-step process so we don't remove
+            // duplicate elements from any of the input iterators.
```
I am not completely certain about the approach taken here. There is some extra complexity to handle iterators producing duplicate elements. The question is whether the invariant expressed by `SortedIterator` means the iterator produces a non-decreasing sequence or is is a strictly increasing sequence. In other words, are the set operations here supposed to handle multisets with duplicate elements or are plain sets sufficient?

Both work options for my use case. Either way I think it would be nice if the docs were more explicit about this.

rklaehn commented:
Awesome. I am currently on holiday and unable to give this a good review, but from first glance this looks good. Will give it a thorough review once I am back.

iljakuklic reviewed (commented):
> Awesome. I am currently on holiday and unable to give this a good review, but from first glance this looks good. Will give it a thorough review once I am back.

Sure, no hurry. Thanks

rklaehn commented:
Hi. Sorry about not following up earlier. It has been a bit crazy because of having my entire family at home due to corona...

So, `SortedIterator` is supposed to produce a strictly increasing sequence, not just non-decreasing. I will update the docs accordingly.

Your code handles that properly, so I guess it is OK.

rklaehn commented on src/sorted_iterator.rs:
```diff
@@ -54,6 +55,94 @@ impl<K: Ord, I: Iterator<Item = K>, J: Iterator<Item = K>> Iterator for Union<I,
 }
 
+// An iterator with the first item pulled out.
+pub(crate) struct Peeked<I: Iterator>{
```
Can you use the normal https://doc.rust-lang.org/std/iter/struct.Peekable.html for this? The only difference seems to be the Reverse. So having just a newtype wrapper for a Peekable might be slightly less boilerplate. Not the end of the world though.

iljakuklic replied:
The problem is that in order to merge the iterators together with good complexity, I use `binary_heap` which requires the elements to implement `Ord`. Therefore, I think a wrapper of one kind or another is still necessary.

There is another subtle difference between the standard `Peekable` and this. The `Peeked` struct defined here always pre-fetches one element out of the wrapped iterator, whereas `Peekable` gets the element on demand. That is again useful to implement `Ord` that compares the iterators by the first element without hacks.

rklaehn replied:
> > Hi. Sorry about not following up earlier. It has been a bit crazy because of having my entire family at home due to corona...
> 
> Hi, hope you and your family are fine.
> 
We are doing well given the circumstances. It is just hard to get some free time for open source work.

> > So, `SortedIterator` is supposed to produce a strictly increasing sequence, not just non-decreasing. I will update the docs accordingly.
> > Your code handles that properly, so I guess it is OK.
> 
> In that case, I will be able to simplify the implementation since handling duplicate elements is not required. Will have a look into it when I have some time over the next couple of days.

Cool. Looking forward to it.

rklaehn replied:
> The problem is that in order to merge the iterators together with good complexity, I use `binary_heap` which requires the elements to implement `Ord`. Therefore, I think a wrapper of one kind or another is still necessary.
> 
Yes, some kind of wrapper is definitely needed.

> There is another subtle difference between the standard `Peekable` and this. The `Peeked` struct defined here always pre-fetches one element out of the wrapped iterator, whereas `Peekable` gets the element on demand. That is again useful to implement `Ord` that compares the iterators by the first element without hacks.

Well, the Peekable is not exactly super complex, so if you think it is the best way I am OK with that.

rklaehn commented on src/lib.rs:
````diff
@@ -161,6 +161,42 @@ pub trait SortedIterator: Iterator + Sized {
+/// );
+/// ```
+pub fn multiway_union<T, I>(iters: T) -> MultiwayUnion<I>
+where I: SortedIterator, T: IntoIterator<Item = I>, I::Item: Ord {
````
This does not enforce at type level that the iterators to be merged are sorted themselves. What happens if one of them is unsorted? Will it properly deal with that?

iljakuklic replied:
Is `I: SortedIterator` not sufficient for that? The function takes a `T`, which is a container full of iterators of type `I`, which is required to be a `SortedIterator`. Am I missing something?

rklaehn replied:
Sorry, I am blind. So it is fine.

iljakuklic commented:
> Hi. Sorry about not following up earlier. It has been a bit crazy because of having my entire family at home due to corona...

Hi, hope you and your family are fine.

> So, `SortedIterator` is supposed to produce a strictly increasing sequence, not just non-decreasing. I will update the docs accordingly.
> 
> Your code handles that properly, so I guess it is OK.

In that case, I will be able to simplify the implementation since handling duplicate elements is not required. Will have a look into it when I have some time over the next couple of days.

rklaehn commented:
> > Hi. Sorry about not following up earlier. It has been a bit crazy because of having my entire family at home due to corona...
> 
> Hi, hope you and your family are fine.
> 
We are doing well given the circumstances. It is just hard to get some free time for open source work.

> > So, `SortedIterator` is supposed to produce a strictly increasing sequence, not just non-decreasing. I will update the docs accordingly.
> > Your code handles that properly, so I guess it is OK.
> 
> In that case, I will be able to simplify the implementation since handling duplicate elements is not required. Will have a look into it when I have some time over the next couple of days.

Cool. Looking forward to it.

iljakuklic committed:
Simplify the multiway union operation

The simplification is possile because handlind of duplicate elements is
not required.

iljakuklic commented:
Just pushed an update simplifying the main union algorithm (the iteratior `next` method). There are no longer any temporary `vec`s. It is in a separate commit for reviewing convenience. Can squash the two commits later if you prefer.
